eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -444,28 +444,28 @@ instance FromCBOR FailureDescription where
       dec 1 = SumD PlutusFailure <! From <! From
       dec n = Invalid n
 
-data ScriptResult = Passes | Fails ![FailureDescription]
+data ScriptResult = Passes ![PlutusDebug] | Fails ![FailureDescription]
   deriving (Show, Generic, NoThunks)
 
 instance ToCBOR ScriptResult where
-  toCBOR Passes = encode $ Sum Passes 0
+  toCBOR (Passes ps) = encode $ Sum Passes 0 !> To ps
   toCBOR (Fails fs) = encode $ Sum Fails 1 !> To fs
 
 instance FromCBOR ScriptResult where
   fromCBOR = decode (Summands "ScriptResult" dec)
     where
-      dec 0 = SumD Passes
+      dec 0 = SumD Passes <! From
       dec 1 = SumD Fails <! From
       dec n = Invalid n
 
-andResult :: ScriptResult -> ScriptResult -> ScriptResult
-andResult Passes Passes = Passes
-andResult Passes ans = ans
-andResult ans Passes = ans
-andResult (Fails xs) (Fails ys) = Fails (xs ++ ys)
-
+-- | The ScriptResult operation is biased towards failure.
```

Comment:
Thanks for bringing this up, it's a bit tricky and I don't know how best to handle it.

If a transaction contains one or more failing scripts, then:
* if the transaction is marked as `IsValid = True`, then the transaction will be rejected. The predicate failure will collect all the failures, and the node logs will display all the needed information to re-run the failed scripts. but indeed there is no mention of the succeeding scripts in this case.
* if the transaction is marked as `IsValid = False`, then as of right now neither the logs nor the events record this information. because of this semigroup instance, it would be easy to make an event for the failures, but an event for all the scripts would need to either re-run everything or we could redo how `evalScripts` works altogether (ie not use `<>`). I'll point out where in this PR we could put the event for the failures that are marked as failures.

Reply:
As a complete outsider, what I would have sort of expected would be that the events would be emitted as a side effect of doing the validation, and orthogonally to how the actual results of validation are reported. But I don't know if that makes any sense here :shrug:

Reply:
it's not quite that simple (at least not without a big refactor), but I can probably make it look more like that from the outside.